### ISSUE TITLE
Has anyone experienced a sudden and drastic increase of the loss while training a Bayesian neural network?

### ISSUE THREAD

**Issue by nbro** [tensorflow/probability#762]
I am trying to train the following Bayesian CNN model, which contains several Bayesian convolution and dense layers. The inputs are images with shape `(64, 80, 3)` and the outputs are normal distributions, one for each of the 289 output neurons. You can see the details of the architecture below.

```
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
camera (InputLayer)             [(None, 64, 80, 3)]  0                                            
__________________________________________________________________________________________________
conv2d_flipout (Conv2DFlipout)  (None, 64, 80, 8)    441         camera[0][0]                     
__________________________________________________________________________________________________
activation (Activation)         (None, 64, 80, 8)    0           conv2d_flipout[0][0]             
__________________________________________________________________________________________________
max_pooling2d (MaxPooling2D)    (None, 32, 40, 8)    0           activation[0][0]                 
__________________________________________________________________________________________________
conv2d_flipout_1 (Conv2DFlipout (None, 32, 40, 16)   2321        max_pooling2d[0][0]              
__________________________________________________________________________________________________
activation_1 (Activation)       (None, 32, 40, 16)   0           conv2d_flipout_1[0][0]           
__________________________________________________________________________________________________
max_pooling2d_1 (MaxPooling2D)  (None, 16, 20, 16)   0           activation_1[0][0]               
__________________________________________________________________________________________________
conv2d_flipout_2 (Conv2DFlipout (None, 16, 20, 32)   9249        max_pooling2d_1[0][0]            
__________________________________________________________________________________________________
activation_2 (Activation)       (None, 16, 20, 32)   0           conv2d_flipout_2[0][0]           
__________________________________________________________________________________________________
max_pooling2d_2 (MaxPooling2D)  (None, 8, 10, 32)    0           activation_2[0][0]               
__________________________________________________________________________________________________
flatten (Flatten)               (None, 2560)         0           max_pooling2d_2[0][0]            
__________________________________________________________________________________________________
dense_flipout (DenseFlipout)    (None, 512)          2621953     flatten[0][0]                    
__________________________________________________________________________________________________
target_locations (DenseFlipout) (None, 289)          296226      dense_flipout[0][0]              
__________________________________________________________________________________________________
target_rhos (DenseFlipout)      (None, 289)          296226      dense_flipout[0][0]              
__________________________________________________________________________________________________
target (DistributionLambda)     ((None, 289), (None, 0           target_locations[0][0]           
                                                                 target_rhos[0][0]                
==================================================================================================
Total params: 3,226,416
Trainable params: 3,226,410
Non-trainable params: 6
__________________________________________________________________________________________________
```

I tried to train this model. The training ELBO loss starts around 61k and the validation ELBO loss around 58k.  Both the training and validation ELBO losses decrease for roughly 10 epochs. However, around the 12th, both the training and validation ELBO losses suddenly and drastically increase. The training loss goes from ~11k to ~27k, while the validation loss, at the 13th epoch, goes from ~7k to ~110k. What could have caused this? Maybe a numerical error? The dataset is quite small: it contains slightly more than 10k instances. However, the non-Bayesian version (i.e. the same layers, but non-Bayesian ones, apart from the output layers, because, in the Bayesian case, I am modelling both the mean and variance of the normal distributions) of this Bayesian model performs a lot better than this Bayesian model.

**Comment by nbro** [comment]
Although I was trying to solve a classification problem with normal distributions as the output, this sudden drop of the loss is still quite mysterious.

**Comment by bgroenks96** [comment]
I think this very likely has to do with the weighting of the KL-divergence, which is likely of higher magnitude than your log likelihood. In my experience, trying to train variational models with completely unweighted KL-divergence almost never works. The variational posterior just immediately converges to the prior, and unless your prior is **really good**, this obviously won't work. This can also happen after several epochs of training.

**Comment by nbro** [comment]
@bgroenks96 I think I was scaling the KL divergence. I was scaling it by 64 (which was the size of my mini-batches).

**Comment by bgroenks96** [comment]
Might be better to scale it as 1/B where B is the *number* of mini-batches.

There is also an argument to be made for scaling as 1/(B*w*h) where w and h are the width and height of the image, i.e. average the KLD over space.
